File: api/app/routers/streets.py
"""Street blocks endpoints."""

from typing import Annotated, Literal, cast

from fastapi import APIRouter, Depends, Query, Request

from app.data_loader import make_refresh_dependency
from app.models.page import Page
from app.utils.pagination import build_page, paginate_features
from dashboard_utils.models.streets import StreetBlockFeature

router = APIRouter(dependencies=[Depends(make_refresh_dependency(["streets"]))])


class StreetsPage(Page):
    """Paginated street blocks GeoJSON response."""

    type: Literal["FeatureCollection"]
    features: list[StreetBlockFeature]


@router.get("/streets")
def get_streets(
    request: Request,
    segment_ids: Annotated[str | None, Query()] = None,
    limit: Annotated[int, Query(ge=0)] = 2000,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> StreetsPage:
    """Return street block GeoJSON filtered by segment IDs.

    Parameters
    ----------
    request : fastapi.Request
        The current request with access to application state.
    segment_ids : str | None, optional
        Comma-separated segment IDs to filter by; if omitted, return all features.
    limit : int, optional
        Maximum number of features to return.
    offset : int, optional
        Zero-based index of the first feature to return.

    Returns
    -------
    StreetsPage
        Paginated GeoJSON FeatureCollection with pagination metadata.
    """
    limit = max(limit, 0)
    # Parse comma-separated segment IDs
    segment_id_list = (
        [sid.strip() for sid in segment_ids.split(",") if sid.strip()] if segment_ids else None
    )
    if not segment_id_list:
        page_features, count, next_offset, total = paginate_features(
            request.app.state.streets_geojson["features"],
            limit=limit,
            offset=offset,
        )
    else:
        features_source = [
            request.app.state.streets_by_segment_id[sid]
            for sid in segment_id_list
            if sid in request.app.state.streets_by_segment_id
        ]
        page_features, count, next_offset, total = paginate_features(
            features_source,
            limit=limit,
            offset=offset,
        )
    return cast(
        StreetsPage,
        build_page(
            features=page_features,
            limit=limit,
            offset=offset,
            count=count,
            total=total,
            next_offset=next_offset,
        ),
    )

File: frontend/src/shared/api/streets.ts
import { apiFetch } from "./client";
import { fetchAllPages } from "@/shared/utils/pagination";
import type { PageMeta } from "@/shared/types/pagination";

/**
 * GeoJSON Feature for street block data.
 */
interface StreetFeature {
  type: "Feature";
  geometry: GeoJSON.Geometry;
  properties: {
    segment_id: string;
    street_name: string;
    block_number: string;
    [key: string]: any;
  };
}

/**
 * Paginated response from the /streets endpoint.
 */
interface StreetsResponse extends PageMeta {
  type: "FeatureCollection";
  features: StreetFeature[];
}

/**
 * Parameters for fetching street data.
 */
interface FetchStreetsParams {
  /** Array of segment IDs to filter by */
  segment_id?: string[];
  /** Maximum number of features to return (default: 2000) */
  limit?: number;
  /** Zero-based offset for pagination (default: 0) */
  offset?: number;
}

/**
 * Fetches a single page of street block GeoJSON data from the API.
 *
 * The API returns paginated results. Use the `next_offset` field in the response
 * to fetch subsequent pages, or use `fetchStreetsAllPages` to fetch all data automatically.
 *
 * @param params - Query parameters for filtering and pagination
 * @returns Promise resolving to paginated street FeatureCollection
 * @throws Error if request fails
 *
 * @example
 * ```typescript
 * // Fetch first page of all streets
 * const page1 = await fetchStreetsPage({ limit: 2000, offset: 0 });
 *
 * // Fetch specific segments
 * const filtered = await fetchStreetsPage({
 *   segment_id: ['123', '456'],
 *   limit: 100,
 * });
 * ```
 */
export async function fetchStreetsPage(
  params: FetchStreetsParams = {}
): Promise<StreetsResponse> {
  const queryParams = new URLSearchParams();

  if (params.segment_id) {
    params.segment_id.forEach((id) => queryParams.append("segment_id", id));
  }
  queryParams.append("limit", String(params.limit ?? 2000));
  queryParams.append("offset", String(params.offset ?? 0));

  const queryString = queryParams.toString();
  return await apiFetch<StreetsResponse>(`/streets?${queryString}`);
}

/**
 * Fetches all street block data by exhausting paginated API endpoints.
 *
 * This function automatically handles pagination by making multiple requests
 * until all data is retrieved. Use with caution for large datasets.
 *
 * @param params - Query parameters for filtering
 * @param params.segment_id - Optional array of segment IDs to filter by
 * @param params.pageSize - Number of features to fetch per request (default: 2000)
 * @returns Promise resolving to complete GeoJSON FeatureCollection with all street features
 * @throws Error if any API request fails
 *
 * @example
 * ```typescript
 * // Fetch all streets
 * const allStreets = await fetchStreetsAllPages();
 * console.log(`Total street segments: ${allStreets.features.length}`);
 *
 * // Fetch specific segments only
 * const filtered = await fetchStreetsAllPages({
 *   segment_id: ['123', '456', '789']
 * });
 *
 * // Fetch with custom page size
 * const allStreets = await fetchStreetsAllPages({ pageSize: 5000 });
 * ```
 */
export async function fetchStreetsAllPages(
  params: {
    segment_id?: string[];
    pageSize?: number;
  } = {}
): Promise<{ type: "FeatureCollection"; features: StreetFeature[] }> {
  return fetchAllPages(
    (paginationParams) =>
      fetchStreetsPage({
        segment_id: params.segment_id,
        ...paginationParams,
      }),
    {},
    params.pageSize
  );
}
